### ISSUE TITLE
Support: Where does AppModel.php learn the expected db_version for updateDatabase

### ISSUE THREAD

**Issue by github-germ** [MISP/MISP#8210]
### Support Questions

Want to understand the db update process better. I see that `AppModel.php` has a big `case` statement in `updateDatabase`, and in the case of say an upgrade from 2.4.148 to 2.4.153 the cases for 73-79 will be invoked as the current 2.4.148 db_version found in the `admin_settings` table is 72.

Here's what I'm missing...
1. Where does the code learn the expected db_version for the new X.Y.Z MISP version being executed?
2. Does it grab it directly from the `db_version` key-value in the `/var/www/MISP/db_schema.json` file, and if so where is that in the PHP? Or if not there, where else does it get that?

Note: I was originaly thinking that the `db_schema.json` file was used to revise the actual schema, but now see that that `case` in `AppModel.php` enumerates through all the updates.

Thanks.

### MISP version

2.4.153

### Operating System

RedHat

### Operating System version

8.4

### PHP version

7.4

### Browser

All

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by righel** [comment]
Hello @github-germ
Here is a breakdown for a common update flow:
1. User logs in / `UsersController:login()`
2. `UsersController::beforeFilter()`
3. `AppController::beforeFilter()`
4. `AppModel::runUpdates()`
The current database version is fetched from the `misp.admin_settings` table:
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L2103-L2106
5. `AppModel::findUpgrades($db_version)` finds out which updates need to be applied, by collecting the versions ahead of the current version in the `AppModel::DB_CHANGES` array:
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L2489-L2493
6. If there are updates available, enqueue a `AdminShell::runUpdates()` background job in the `prio` queue to apply them:
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L2172-L2181
7. The background job executes `AdminShell::runUpdates()`, who then calls `Server::runUpdates()` with `$useWorker=false`
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Console/Command/AdminShell.php#L501
8. Then `AppModel::runUpdates()` is called again, steps 4-5 are done again, but this time since `$useWorker=false` the task is executed, one update at a time:
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L2221
9. `AppModel::updateMISP` handles some specific version updates, where some actions besides DB updates are required, otherwise the default case handles it:
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L227
10. Finally `AppModel::updateDatabase($command)` runs each update.
https://github.com/MISP/MISP/blob/ab1305cc18ef23a8d0c2d7d354d6fc8901ab19ad/app/Model/AppModel.php#L1680-L1683
   * NOTE: Step 9-10 is looped for each pending update until the system is up-to-date

Regarding `db_schema.json` is file is generated manually on each release via `AdminShell::dumpCurrentDatabaseSchema()` and only used in the `Diagnostics` page, `SQL database status` / `Schema status` section.

Let me know if this answers your questions.

**Comment by github-germ** [comment]
Perfect. Many thanks. I missed your step 5 above. Two follow-up questions:
1. Re' `DB_CHANGES`...
* When I look, for example, at the span from `72` to `79` (e.g. for 2.4.148 -> 2.4.153) why are so many set `false`, i.e. why were the updates added to the  big `case` block in `updateDatabase` but then not executed?
2. Re' `db_schema.json` ...
* Why does it appear that `db_schema.json` does not match our actual schema with some settings in that file having no matching update in `updateDatabase`?
* We found this quite confusing when trying to validate our schema. And if this mismatch is true, it seems like WebUI Diagnostics is misleading?

**Comment by righel** [comment]
1. It was a mistake, the `false/true` in the `AppModel::DB_CHANGES` array means that the given version requires a mass logout of the users, updated the previous comment to avoid future confusions.
2. This diagnostic could in fact be misleading, in a couple of scenarios such as:
   * Making manual changes in your database, such as new indexes/other.
   * Having a database engine/version that differs from the one used by us when generating the `db_schema.json` file via `AdminShell::dumpCurrentDatabaseSchema()`.
   * A release was made without properly updating the `db_schema.json` file.

**Comment by github-germ** [comment]
1. OK, yea... I see `DB_CHANGES` is for logout. So indeed in my example all the cases up to `79` did run.
2. I'm still confused about `db_schema.json` delivered with the release as it has schema settings that I cannot find in the `updateDatabase` directives in the `case` statement.

**Comment by iglocska** [comment]
@righel is spot on with the explanation of the flow. DB changes has 2 separate functionalities, the keys are indeed the list of scripts to execute and the values (true/false) just indicate the mass logout. All cases up to 79 must have been executed for you, we basically only proceed to the next update once the previous one has been successfuly applied.

For the schema: It's as @righel explains, simply a valid JSON schema dropped by a developer before a release. Last time I did it, I spun up a new MISP instance, ingested the MYSQL.sql initial sql script to create the baseline table structure, then on the first logon MISP will execute all the updates that have been made since the creation of the MYSQL.sql file's contents (which also contains a DB version). 

So basically, db_schema.json should reflect a pristine fresh installation of MISP with all updates applied.

**Comment by packet-rat** [comment]
_Editorial comment:  Thanks for the timely and effective responses!  They are much appreciated._
